container/build.py

```diff
@@ -0,0 +1,156 @@
+}
+
+
+def main(type_, project, container, tag, async_, bucket, local):
+    project_root = subprocess.check_output(
+        "git rev-parse --show-toplevel", shell=True).strip()
+    package = TYPE_PACKAGE_MAP[type_]
+    target = TYPE_TARGET_MAP[type_]
+    tarball = TYPE_TARBALL_MAP[type_]
+
+    # We need to start the build from the root of the project, so that we can
+    # mount the full root directory (to use bazel builder properly).
+    os.chdir(project_root)
+    # We need to run clean to make sure we don't mount local build outputs
+    subprocess.call("bazel clean", shell=True)
+
+    if local:
+        print("Building container locally.")
+        subprocess.call(
+            "bazel run //{}:{}".format(package, target), shell=True)
+        print("Testing container locally.")
+        subprocess.call(
+            "bazel test //{}:{}-test".format(package, target), shell=True)
+        print("Tagging container.")
+        subprocess.call(
+            "docker tag bazel/{}:{} {}:latest".format(package, tarball, type_))
+        print(("\n" +
+               "{TYPE}:lastest container is now available to use.\n" +
+               "To try it: docker run -it {TYPE}:latest \n").format(TYPE=type_)
+              )
+    else:
+        print("Building container in Google Cloud Container Builder.")
+        # Setup GCP project id for the build
+        subprocess.call(
+            "gcloud config set project {}".format(project), shell=True)
+        # Ensure all BUILD files under /third_party have the right permission.
+        # This is because in some systems the BUILD files under /third_party (after git clone)
+        # will be with permission 640 and the build will fail in Container Builder.
+        for dirpath, _, files in os.walk(project_root+'/third_party'):
+            for f in files:
+                full_path = os.path.join(dirpath, f)
+                os.chmod(full_path, 0o644)
+
+        config_file = "{}/container/cloudbuild.yaml".format(project_root)
+        extra_substitution = ",_BUCKET={},_TARBALL={}".format(bucket, tarball)
+        if not bucket:
+            config_file = "{}/container/cloudbuild_no_bucket.yaml".format(
+                project_root)
+            extra_substitution = ""
+        ASYNC = ""
+        if async_:
+            ASYNC = "--async"
+        subprocess.call(
+            "gcloud container builds submit . \
+            --config={CONFIG} \
+            --substitutions _PROJECT={PROJECT},_CONTAINER={CONTAINER},_TAG={TAG},_PACKAGE={PACKAGE},_TARGET={TARGET}{EXTRA_SUBSTITUTION} \
+            --machine-type=n1-highcpu-32 \
+            {ASYNC}".format(
+            CONFIG=config_file,
+            PROJECT=project,
+            CONTAINER=container,
+            TAG=tag,
+            PACKAGE=package,
+            TARGET=target,
+            EXTRA_SUBSTITUTION=extra_substitution,
+            ASYNC=ASYNC),
+        shell=True)
+
+
+if __name__ == "__main__":
```

Comment:
Could we move the parameter parsing to its own function, and add it to `main`?

It is a common practice to have less logic here and have this block only calls main(), like
``` python
if __name__ == "__main__":
    main()
```

Reply:
Moved argument parsing to a function. The if block now just gets the args and passes them to main